### DISCUSSION TITLE
cli - supabase db push as postgres user causes "ERROR: must be owner of table"

### DISCUSSION THREAD

**Discussion by mikelyndon** [supabase/supabase#6326] [Questions] [answered]
In an attempt to push a migration from local to remote I'm getting the error
`Error: ERROR: must be owner of table <table> (SQLSTATE 42501)`
because the table is owned by `supabase_admin`

I've tried running, `GRANT ALL ON TABLE public.<table> TO postgres;` from the dashboard but still no luck.

So my next thought was to change the cli user from `postgres` to `supabase_admin`.

Is there any reason why this is a bad idea?

**Comment by cmantsch** [comment]
I also get a similar error when trying to run the migrations from cli on a project hosted by supabase:
`Error: ERROR: must be member of role "anon" (SQLSTATE 42501)`

Setting the postgres user temporarily to superadmin as described in [this ticket](https://github.com/supabase/supabase/issues/1416) fixes it, but this doesn't feel like the proper solution. On a self-hosted supabase instance the migrations run just fine.

**Reply by mupinnn** [reply]
Did you change the role of the table from Supabase Studio SQL editor locally? I got the same error when trying to push my local migrations to the remote.
```sh
supabase db push # sudo supabase db push
Applying unapplied migrations...
Error: ERROR: must be owner of table profile (SQLSTATE 42501)
```
I tried to alter the postgres user to superadmin with `ALTER USER postgres WITH SUPERUSER` , then I ran `supabase db push` again and the same error came up. Hereafter, I tried again to alter the user to superadmin then I ran `supabase db remote commit` first and next `db push` still resulting the same error. Am I missing some step here?
<br />
> On a self-hosted supabase instance the migrations run just fine.

@cmantsch The self-hosted instance can use the CLI to? I might move from the CLI to the self-hosted instead

**Reply by cmantsch** [reply]
Totally missed your reply @mupinnn, sorry!

It's been a while since I used the CLI, but if I remember correctly, you can set a URL of the instance for applying the migrations. We built a rudimentary docker container for running migrations like that in our CI

**Comment by klufkin** [comment]
I've been running into the same issue when trying to push my local migrations up.
```
supabase db push
Applying unapplied migrations...
Error: ERROR: must be owner of table profile (SQLSTATE 42501)
```

I can confirm that @cmantsch's solution works in getting around this issue. Hopefully a longer lasting solution will arise!

**Comment by mikelyndon** [comment, marked as answer]
As @cmantsch mentioned, the current solution is setting the postgres user temporarily to superadmin. These steps are also listed in the [guide to migrate between projects](https://supabase.com/docs/guides/database#migrating-between-projects).
In the SQL Editor on the **remote** dashboard run `ALTER ROLE postgres SUPERUSER`
Then run `supabase db push` and finally in the remote SQL Editor run `ALTER ROLE postgres NOSUPERUSER`

You can also change the owner of the table `ALTER TABLE public.<table> OWNER TO "postgres";`

Yes, this is a workaround, but hopefully this is ironed out in the near future.

**Reply by thomasmol** [reply]
Thanks for the workaround, it works for me!
Do you have any ideas updates on a more permanent fix?

**Reply by JSmithOner** [reply]
@mikelyndon do you know if this SQL command can be ran inside a CI pipeline? this would avoid having to do this operation on each push. Thanks in advance

**Reply by mikelyndon** [reply]
I think this is the official response - change permissions for `postgres` permanently in the dashboard.
https://supabase.com/docs/guides/cli/cicd-workflow#permission-denied-on-db-push

**Reply by JSmithOner** [reply]
thanks @mikelyndon

**Reply by henningko** [reply]
@mikelyndon Unfortunately, the response has changed to changing owners of the table to `postgres`:
https://supabase.com/docs/guides/cli/managing-environments#permission-denied-on-db-push

However, this is an incomplete fix, since it doesn’t work for operations on tables created through `CREATE EXTENSION`, see:
https://github.com/supabase/cli/issues/647

I just hope the `supabase_admin` vs `postgres` confusion gets fixed altogether.

**Reply by thellimist** [reply]
`ALTER ROLE postgres SUPERUSER`

This gives `ERROR:  42501: permission denied to alter role DETAIL:  Only superusers can alter privileged roles.` which means this hack doesn't work on the hosted dashboard.

Any solutions?
